### ISSUE TITLE
MockWebServer: Clear requests before each test

### ISSUE THREAD

**Issue by GabrielBB** [square/okhttp#6212]
Can we have a method to clear requests? Something like this:

```kotlin
    @BeforeEach
    fun cleanRequests() {
        mockServer.clearRequests();
    }
```

My tests are taking requests from previous tests. I wouldn't want to  shutdown entirely

**Comment by JakeWharton** [comment]
Are you using the test rule? There should be a new instance for every test.

**Comment by GabrielBB** [comment]
Yes, I'm using `@TestInstance(TestInstance.Lifecycle.PER_CLASS)`

**Comment by JakeWharton** [comment]
Can you switch to PER_METHOD? Assuming such a thing exists.

**Comment by swankjesse** [comment]
Yeah, MockWebServer instances are not expected to be reused.

**Comment by GabrielBB** [comment]
Will a clear method be enough to be reusable ? I can open PR, why not I guess ?

**Comment by swankjesse** [comment]
No, I don't think it's a good fit for a PR. The clear method encourages a pattern we don't support well.

**Comment by nightswimmings** [comment]
What if we need  an static MockWebServer in order to register @DynamicPropertySource properties with the mockWebServer.getPort()

**Comment by yschimke** [comment]
Can't you change your logic to update the DynamicPropertyRegistry as each test runs with the new port/url?

**Comment by nightswimmings** [comment]
DynamicPropertyRegistry method must be static as per documentation, so I guess it does not reevaluate within the class. I think that would imply restarting the whole spring context since the mechanism is similar to that of @SpringBootTest(properties=X). So we'd need the mockrestserver instance to be static as well, I am not sure it makes sense to reinstantiate it globally and start/shutwoning on each test, I find the clearRequests() approach 
more comfy for that scenario

**Comment by yschimke** [comment]
The method is static, but it provides a Supplier that gets called on demand.  I'm not saying it's ideal for you, just that it doesn't appear that you are blocked as you could update the port each time a new MockWebServer is started for your test.

**Comment by nightswimmings** [comment]
I don't understand the Supplier part but I am interested.. :)

I have this currently, how would you exactly do it?

```
static MockWebServer mockWebServer;

    @DynamicPropertySource
    static void properties(DynamicPropertyRegistry propsRegistry) {
         propsRegistry.add("spring.cloud.gateway.routes[0].uri", () -> "http://localhost:" + mockWebServer.getPort())
    }
    
    @BeforeAll
    static void beforeAll() throws IOException {
        mockWebServer = new MockWebServer();
        mockWebServer.start();
    }

    @AfterAll
    static void afterAll() throws IOException {
        mockWebServer.shutdown();
    }
```

**Comment by yschimke** [comment]
Maybe something like the following, where port should be read everytime the property is accessed.

```
     MockWebServer mockWebServer;
     static int port;

    @DynamicPropertySource
    static void properties(DynamicPropertyRegistry propsRegistry) {
         propsRegistry.add("spring.cloud.gateway.routes[0].uri", () -> "http://localhost:" + port)
    }
    
    @Before
    void before() throws IOException {
        mockWebServer = new MockWebServer();
        mockWebServer.start();
    }

    @After
    void afterAll() throws IOException {
        mockWebServer.shutdown();
    }

```

**Comment by nightswimmings** [comment]
Ahh I think I understand you. @DynamicPropertySource is only evaluated once, but by your words I understand that the property supplier itself does it every time the property is accessed.  If so I guess your example is missing a port = mockWebServer.getPort() at the end of the before(), right?

**Comment by nightswimmings** [comment]
@yschimke  I did some tests and the evaluation for the supplier is done once at context startup and that occurs before the BeforeEach, so even if we managed to tell spring to reevaluate whole context before each test, I should make sure that the BeforeEach is still executed before the property evaluation, otherwise I need to do an ugly hack for reinitializing the mockwebserver on the supplier only as long as it takes part during the same test

**Comment by yschimke** [comment]
Is it possible the item reading the property is in the wrong scope? Singleton?

https://www.baeldung.com/spring-dynamicpropertysource

> If our PostgreSQL container is going to listen to a random port every time, then we should somehow set and change the spring.datasource.url configuration property dynamically. Basically, each test should have its own version of that configuration property.

**Comment by tberger-plugsurfing** [comment]
May I ask if there is any solution here? Is okttp Mockserver just unusable with Spring tests and their lifecycle when a @DynamicPropertiesSource is used in combination with a static instance? I have the same problem as stated above and I actually don't understand why the queue can't just be cleared via API.

**Comment by anatoly-spb** [comment]
As a workaround we can find random port before all tests and use it for all MockWebServer instances:
```
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles(profiles = { "test" })
@EnableAutoConfiguration
class ApiServerTests {
  static final String mockBackEndHost = "localhost";
  static final int mockBackEndPort = TestSocketUtils.findAvailableTcpPort();

  @DynamicPropertySource
  static void setupConfig(DynamicPropertyRegistry r) {
    //
    // back end
    //
    r.add("api.server.url", () -> {
      return String.format("http://%s:%s/api", mockBackEndHost, mockBackEndPort);
    });

    //
    // keycloak
    //
    r.add("spring.security.oauth2.resourceserver.jwt.issuer-uri", () -> {
      String uri = keycloak.getAuthServerUrl() + "/realms/rolis";
      return uri;
    });
  }

  MockWebServer mockBackEndServer;

  @BeforeEach
  void setUp() throws IOException {
    mockBackEndServer= new MockWebServer();
    mockBackEndServer.start(InetAddress.getByName(mockBackEndHost), mockBackEndPort);
  }

  @AfterEach
  void tearDown() {
    try {
      mockBackEndServer.close();
    } catch(Exception e) {
    }
  }
  ...
}
```

As the result we reach our goals: 
1. use clean MockWebServer instance for each test
2. use DynamicPropertySource for context configuration

**Comment by yschimke** [comment]
closing as won't fix

last maintainer discussion in Jan 2024, so it's not happening.
